### ISSUE TITLE
Performance: Improve rendering time of the Page (Discovery)

### ISSUE THREAD

**Issue by miina** [GoogleForCreators/web-stories-wp#8255]
## Feature Description
This needs some discovery for detecting ways to improve the page rendering speed.

From the initial render investigation:
Using `React.Profiler` it seemed like rendering the `DisplayLayer` took approximately the same time (or sometimes even less) to render as the `FramesLayer`. E.g. in one run, the display took 360ms and the frames layer took 480ms.

- `FramesLayer`, however, is not really used until the user interacts with the editor. Perhaps there's a way to optimise this so that we wouldn't need to render the frames all the time. E.g. maybe it would only make sense displaying for the active elements and when the user is hovering/focused on the page?

- Additionally, in case of a background image, it seems to always use the fullsize image in the editor, no matter the screen size. Is that purposefully done so?

- Related : #8254 -- perhaps we can also use a placeholder for the page when initially rendering to reduce the jumpiness / offer a better user experience.

It would be good to spend some time to investigate what else can be done to reduce the render time of a Page.

## Additional Context

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

**Comment by Trapsta** [comment]
## Page Loading performance Discovery

For this discovery, I wanted to focus more on the story-editor in isolation and ignore any other bottlenecks that might be unrelated as can be seen in this Lighthouse audit report.

  

![](https://user-images.githubusercontent.com/276316/141190896-9873793e-2c4c-43e4-93e1-91fe8c730750.png)

[Lighthouse Report - Nov 11 2021.zip ](https://github.com/Trapsta/casino-app/files/7516152/localhost_8899-20211110T160636-first-run.html.zip)

  

### Test Lab Environment

1. Network Throttling - Fast 3G

2. CPU Throttling - None

3. Disable caching

4. Incognito session

5. Development build - ( would be better to use the prod build but some tools for profiling have been stripped to optimize the bundle 🥲)

  

### Test Lab Report

  

Saving the chrome dev tools performance profile report here to benchmark any future improvements. Profile is in json format and is easily portable and shareable.

  

[perf profile Nov 11 2021.json.zip](https://github.com/Trapsta/casino-app/files/7516585/Profile-20211110T175141-local-optimized-first-run.json.zip)

  

To optimize the test lab for the editor, I added a button that when clicked loads the editor instead of starting profiling with page reload to ignore any non-editor related bottlenecks.

  

The most trouble looks like it's coming from javascript and main thread work as seen in this screenshot.

  

![](https://user-images.githubusercontent.com/276316/141201578-47e67fd3-1df1-4046-bbf8-81303699511d.png)

  

The CPU activity bar color codes the different operations running while the editor is starting up:

  

- Grey - **Loading**: Making network requests and parsing HTML

- Yellow - **Scripting**: Parsing, compiling, and running JavaScript code, also includes Garbage Collection (GC)

- Purple - **Rendering**: Style and layout calculations

- Green - **Painting**: Painting, compositing, resizing and decoding images

  

One particular cuplrint here is the `getImageMetaData` function from the color thief library. It runs a lot of times before the page renders and takes a lot of time to execute.

  

![](https://user-images.githubusercontent.com/276316/141203323-2a9b6d6c-ff15-4115-9fcf-3e4fdd85eaf1.png)

  

It also causes the browser to repaint everytime it runs as seen in the trace below This is an expensive operation that we should definitely avoid.

  

![getImageData](https://user-images.githubusercontent.com/276316/141204348-77c8442f-081a-47ff-8209-aa8a8abd55a4.jpg)

  

Hopefully, [#9550](https://github.com/google/web-stories-wp/issues/9550) which is in progress can help here otherwise we might need to look into more strategies to help with this.

  

We should also find other ways of reducing the time the browser takes for paint and rendering operations.

  

![](https://user-images.githubusercontent.com/276316/141207649-5deacb6c-1ccd-4be5-a6d7-1a62202ddaed.png)

  

The screencast below flashes green whenever the browser needs to repaint. It seems like it does this quite a number of times before the page is fully loaded.

  

![](https://user-images.githubusercontent.com/276316/141210655-c4e24a01-8b95-4072-9996-3c55f19236e5.gif)

  

Found this great guide to simplifying and optimizing paint complexity https://developers.google.com/web/fundamentals/performance/rendering/simplify-paint-complexity-and-reduce-paint-areas and I think some of the strategies here might help with this.

  

Wasteful rerenders also seem to be slowing the page down. Components seem to be rendering way too frequently. From the profiler screenshot below, there are 21 commits to the DOM even before any user interaction.

  

![](https://user-images.githubusercontent.com/276316/141209189-baf2d915-8b95-45a0-9d3b-cdd9f15d5560.png)

  

This is probably related to the general architecture of the editor where some parents cause some child nodes to rerender unnecessarily. Restructuring the components could help or like Morten suggested bringing in a state management toolset like [Redux toolkit].(https://redux-toolkit.js.org/) to make the react components re render as less frequently as possible.

  

The profiler also highlights some components which take a lot of time to render. Optimizing these sluggish components could help reduce the page load time.

![](https://user-images.githubusercontent.com/276316/141209578-3aa8b6e4-719a-4c88-9e55-b03c5af69886.png)

  

![](https://user-images.githubusercontent.com/276316/141209649-941ab6dd-8ede-4053-8efb-8ad0c75f7e3b.png)

 Prioritizing the `DisplayLayer` could also help speed up the page load. I tested Miina's theory that not rendering the `FramesLayer` initially shows a loading performance  boost as illustrated below.

![](https://user-images.githubusercontent.com/276316/141327150-07dad84d-9434-4a60-a67b-ae91999a448e.png)
Page starts rendering at around the 5700ms mark with the `FramesLayer` enabled


![](https://user-images.githubusercontent.com/276316/141328299-c07984c4-9cea-4af5-9cf9-917665642902.png)
Page starts rendering at around the 3200ms mark with the `FramesLayer` enabled

Another strategy we can use to improve the page rendering is optimizing loading of assets mostly media. As shown on the screenshot below, even though the assets are loading in parallel , they take some time to request. Maybe caching helps here but I only tested with caching disabled , if not we can load them after the page assets have been loaded and the page has finished rendering.

  

![](https://user-images.githubusercontent.com/276316/141205829-dfeea6ee-42b3-4c38-9516-a5fd0d803530.png)


### Summary
In conclusion the next steps would be to:

 -  Avoid forced reflows by optimizing use of the color handling library.
 - Prioritize rendering the `DisplayLayer` over the `FramesLayer` in initial load.
 - Optimize the `Button`  , `TextDisplay` and `AnimationContainer` components which seem to be slow in the initial render.
 - Loading media in the library after the page has fully loaded to maximize idle time.
 - Look into general app architecture and state management toolsets to reduce wasteful rerenders in child nodes when parent nodes props or state changes.

**Comment by miina** [comment]
@Trapsta Thanks for looking into this. (Ping @barklund as well -- requesting for review. EDIT: saw later that you added some notes in Slack already)

Some additional questions:
> Prioritize rendering the DisplayLayer over the FramesLayer in initial load.

- Was the check you did for the full `FramesLayer`? The `FramesLayer` also includes the Page navigation arrows and the Page Menu. We'd still need to render those all the time but I'm wondering about the Frames of the elements -- these are not needed / visible until the user interacts with the editor, so not displaying the element frames initially might be something we could perhaps benefit from, e.g. displaying the Frame of the element only when actually needed (hovering/when selected/drop-targeting, etc.) -- would that make a difference, or is it only relevant with full `FramesLayer`?

> Optimize the Button , TextDisplay and AnimationContainer components which seem to be slow in the initial render.

- Were these specifically the buttons of the Page Menu or the quick-action menu or something else? Were some specific buttons rendering slow, or all?

- Did you also have a chance to look into re-rendering a page (more than just the initial load)?

**Comment by Trapsta** [comment]
@miina I initially tested without the entire `FramesLayer` but even after testing with only the page's `FrameElement`s not rendered, I can see a performance gain for both initial rendering and navigating between pages.

Without the `FrameElement`s, page starts rendering at 3200ms and completes rendering earlier.
![Screenshot 2021-11-15 at 15 50 13](https://user-images.githubusercontent.com/276316/141785007-dacc6578-ade7-44ba-a9d6-e407b1e9957e.png)

With the pages `FrameElement`s , page starts rendering at around 4200ms and takes more time to completely render
![Screenshot 2021-11-15 at 15 49 42](https://user-images.githubusercontent.com/276316/141785405-c5890032-5fbc-4508-90c6-36f68f50c1bf.png)

In regards to the button component, the button that was taking a lot. of time to render was the help center (tips) button.

I have tested navigating between pages apart from the initial rendering but I only tested the impact of the `FrameLayer` as laid out above.

**Comment by miina** [comment]
@Trapsta Thanks for the additional details. Mentioned it in Slack as well but it would be great to have the specific comparison times for rendering a Page with `FrameElement`s vs without (e.g. 500ms vs 300ms). The starting time is great to know as well but would be good to compare the specific times as well.

Could you also write the time in ms that the Help Center button took for rendering as a reference?

Once you add these details create the follow-up issues for this discovery (please assign those to me), we should be able to close this issue.

**Comment by Trapsta** [comment]
Sure thing @miina. These are the timings with the network throttled to fast 3G and no cpu throttling. The page is fully  rendered at the 8500ms mark when rendered without the `FrameElement` and when rendered along with the `FrameElement` , completes rendering at 9000ms. The Help Center button took 71.3ms although on subsequent runs it seems to take way less so could be a false positive 
![Screenshot 2021-11-17 at 13 57 02](https://user-images.githubusercontent.com/276316/142188245-b07066ee-e6db-492d-b67c-543d7107410b.png)

**Comment by miina** [comment]
Follow-up issues:
#9727 
#9728 
#9729 

Note: this discovery was mainly focusing on rendering the Page as part of the initial load of the editor. Additional performance wins can be investigated when looking into the Grid performance issues.

**Comment by miina** [comment]
Closing to be followed up in the issues mentioned above and @Trapsta will follow-up the additional questions under those issues.
